app/controllers/api/stages_controller.rb

```diff
@@ -0,0 +1,42 @@
+class Api::StagesController < Api::BaseController
+  include CurrentProject
+
+  before_action :require_project
+
+  def index
+    render json: @project.stages
+  end
+
+  def clone
+    new_stage = Stage.build_clone(old_stage)
+    new_stage.name = stage_name
+    if new_stage.save
+      if deploy_group_ids.any? && deploy_groups.any?
+        new_stage.deploy_groups = deploy_groups
+        new_stage.save
+      end
+
+      render json: new_stage, status: 201
+    else
+      render json: new_stage.errors, status: 422
```

Comment:
yeah ... was thinking that we might need this general rescue workflow
anyway ... either way is fine

On Tue, Aug 9, 2016 at 11:06 AM, Adam Panzer <notifications@github.com>
wrote:

> In app/controllers/api/stages_controller.rb
> <https://github.com/zendesk/samson/pull/1163#discussion_r74112565>:
>
> > +  def index
> > +    render json: @project.stages
> > +  end
> > +
> > +  def clone
> > +    new_stage = Stage.build_clone(old_stage)
> > +    new_stage.name = stage_name
> > +    if new_stage.save
> > +      if deploy_group_ids.any? && deploy_groups.any?
> > +        new_stage.deploy_groups = deploy_groups
> > +        new_stage.save
> > +      end
> > +
> > +      render json: new_stage, status: 201
> > +    else
> > +      render json: new_stage.errors, status: 422
>
> why? that seems more expensive. I've always been told that rescuing costs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1163/files/ec9176e57ff6e9773ea877bdef9761f323850b46#r74112565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ5-3oled_yRAOWXL_Jf4NfSj1H00ks5qeMGggaJpZM4JVi_g>
> .
>